table_test.go

```diff
@@ -30,6 +30,14 @@ type testOutput struct {
 	t *testing.T
 }
 
+type tIteratorProvider struct {
```

Comment:
I would have expected something like a typed function here instead. Something like:

```
type iteratorProviderFunc func(arrow.Record) error

func (f iteratorProviderFunc) Iterator() func(arrow.Record) error {
  return f
}
```

Reply:
that would make it cleaner, I'll make that change

Reply:
@brancz I think that makes sense wrt to the options to sync the query, I'll remove that.

I think for the query that's posted above, could we not execute the filter of the aggregate results in parallel?
```golang
TableScan("table).
Filter(Col("xyz").Regex("abc.*")).
Aggregate("sum", "value", "xyz").
Filter(Col("sum(value)").LT(0))
```

```
       A           B               C               D             E                   F              G               H
=====================================================================================================================================     
┌───────────┐ ┌───────────┐                                                     ┌───────────┐  ┌───────────┐  ┌───────────┐
│           │ │           │                                                     │           │  │           │  │           │
│ Finisher  ├─┤ Filter    ├─┐                                                ┌──┤ Hash      │◄─┤ Filter    │◄─┤ Table Scan│
│           │ │           │ │                                                │  │ Aggregate │  │           │  │           │
└───────────┘ └───────────┘ │                                                │  └───────────┘  └───────────┘  └───────────┘
                            │                                                │
┌───────────┐ ┌───────────┐ │ ┌───────────┐  ┌───────────┐  ┌───────────┐    │  ┌───────────┐  ┌───────────┐  ┌───────────┐
│           │ │           │ │ │           │  │ Concurrent│  │           │    │  │           │  │           │  │           │
│ Finisher  ├─┤ Filter    ├─┼─┤ Exchange  │◄─┤ Hash      │◄─┤ Merger    │◄───┼──┤ Hash      │◄─┤ Filter    │◄─┤ Table Scan│
│           │ │           │ │ │           │  │ Aggregate │  │           │    │  │ Aggregate │  │           │  │           │
└───────────┘ └───────────┘ │ └───────────┘  └───────────┘  └───────────┘    │  └───────────┘  └───────────┘  └───────────┘
                            │                                                │
┌───────────┐ ┌───────────┐ │                                                │  ┌───────────┐  ┌───────────┐  ┌───────────┐
│           │ │           │ │                                                │  │           │  │           │  │           │
│ Finisher  ├─┤ Filter    ├─┘                                                └──┤ Hash      │◄─┤ Filter    │◄─┤ Table Scan│
│           │ │           │                                                     │ Aggregate │  │           │  │           │
└───────────┘ └───────────┘                                                     └───────────┘  └───────────┘  └───────────┘
```

I think the interesting challenge we'll have is coordinating when each step should call the next callback. For example, we wouldn't want to have step D (Concurrent Hash Aggregate), call it's next callback to pass the results on step C (Exchange) until it has seen every record that will be emitted by step E (Merger).

I think the concept of a Finish/Next Finish function on each step could be used for this. The Finish function can be called on physical plan when all records have been sent into the iterator. Steps that want to wait to emit results to the next step will call their nextCallback in their Finish function.

It might also be a strategy that's more maintainable than having the physical plan building one monolithic Finish.

Here's an example (pseudocode) of how this could work:

```golang
type PhysicalPlan interface {
    // NEW:
    Finish() error
    SetNextFinish(next func() error)
    // ...
}
```

```golang
Filter:
  Callback(record) error {
    // filter records
    nextCallback(record)
  }
  Finish() error {
    // filter doesn't have any special finish logic
    return nextFinish()
  }
```

```golang
HashAggregate:
    Callback(record) error {
        // same as current logic
    }
    Finish() error {
        // same as current logic.
        // then do ...
        return nextFinish()
    }

```

```golang
// Merger takes the results from multiple parallel streams and emits them serially
Merger:
    // waiting for all the previous steps to call Finish. Initialized in builder
    wg sync.WaitGroup{}
    
    // only one thread will call the next callback
    mtx sync.Mutex{} 

    Callback(record) error {
        // buffer record
        return nil
    }

    // this will get called from multiple threads, but only one thread waits
    Finish() error {
        wg.Done() // reduce wg w/ each call

        // if this is the waiting thread, wait for all records buffered then emit
        if ok := mtx.TryLock(); ok {
            defer mtx.Unlock();
            wg.Wait()
            for _, record := buffer {
                nextCallback(record) // return err if not nil etc. 
            }
        }
    }
```

```golang
// Exchange takes a serial stream of results and emits the concurrently
Exchange {
    eg ErrorGroup
    rgChan chan arrow.Record

    Callback(record) error {
        rgChan <- record
        return nil
    }

    SetNextCallback (cb) {
        eg.Go(func() error {
            for rec <- rgChan {
                cb(rec) // return err if not nil
            }
        })
    }

    Finish() error {
        close(rgChan)
        eg.Wait() // return error if not nil
        nextCallback()
    }
}
```

Reply:
I think the overall strategy is sound here, I agree it's still pretty akin to the exchange operator, but essentially exchange operator in a map/reduce style system. Something we haven't talked about too much is that we should be able to produce the graphs we keep drawing by hand as a result of the physical planning, so we can inspect what our software decided to do. We don't have to do this as part of this PR, but it's something we need to keep in mind.

Reply:
Sounds good @brancz. Given that we want to draw the plan, maybe it makes sense for each step in the plan to have a reference to the next step itself, as opposed just some of it's methods. What do you think about that? 

i.e. instead of doing something:
```golang
// in the builder
phyPlan.SetNextCallback(prev.Callback)
phyPlan.SetNextFinish(prev.Finish)
```
... we do something like this:
```golang
// change the interface so plans reference the next physical plan
type PhysicalPlan interface {
	Callback(r arrow.Record) error
	Finish() error
	SetNextPlan(PhyisicalPlan)
}

// in the builder:
phyPlan.SetNextPlan(prev)
```
What's nice about doing it like this is that if we extend the PhysicalPlan interface with some Draw() method, it could be called recursively as we traverse the plan and build up the picture.

Reply:
Yeah, I like that!